### PR TITLE
GDB-10752 Add password hash for the security provisioner

### PR DESCRIPTION
**Pull request by cloudvlad**
The password hash is introduced due to the dynamic nature of the bcrypt algorithm, which generates a new hash each time deployment is initiated.

This resolves potential issues with CI/CD tools that detect these changes and prevent false-positive drift events.

Issue: [GDB-10752](https://ontotext.atlassian.net/browse/GDB-10752)

[GDB-10752]: https://ontotext.atlassian.net/browse/GDB-10752
#118

### CODE REVIEW

File: files/config/users.js [hunk 1]
```diff
@@ -16,7 +16,7 @@
     {{- include "grahdb.security.extra-users.json" . | nindent 4 }}
     "{{ .Values.security.provisioner.username }}" : {
       "username" : "{{ .Values.security.provisioner.username }}",
-      "password" : "{bcrypt}{{ htpasswd .Values.security.provisioner.username .Values.security.provisioner.password | trimPrefix (printf "%s:" .Values.security.provisioner.username) }}",
+      "password" : {{ .Values.security.provisioner.passwordHash | default (printf "{bcrypt}%s" ( htpasswd .Values.security.provisioner.username .Values.security.provisioner.password | trimPrefix (printf "%s:" .Values.security.provisioner.username))) | quote }},
```
Comment (mihailradkov) [inline files/config/users.js:19]: Could you try extracting the bcrypt-ing in a helper template function?

File: values.yaml [hunk 2]
```diff
@@ -257,6 +257,9 @@ security:
     # Note: The user is created regardless of if the existingSecret is provided or not.
     username: provisioner
     password: iHaveSuperpowers
+    # Bcrypt-hashed password
```
Comment (mihailradkov) [inline values.yaml:260]: Could expand the description a bit 

> Bcrypt-hashed password for the provisioner user.

Also, see that I've mentioned the format in `security.admin.initialPassword`. Could do that here as well

Comment (mihailradkov) [inline values.yaml:260]: Update the changelog

**Comment by mihailradkov** [comment]
This change should be in the 11.2 feature branch.

File: values.yaml (hunk 2 above)
Comment (cloudvlad) [reply values.yaml]: done

Comment (cloudvlad) [reply values.yaml]: done

File: files/config/users.js (hunk 1 above)
Comment (cloudvlad) [reply files/config/users.js:19]: done

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,8 +10,9 @@
     - Local backups support saving the GraphDB backup archives in volume from an existing persistent volume claim, configured
       with `backup.local`
     - Cloud backups support uploading the GraphDB backup archives in one of the supported cloud object storage services, configured
-      with `backup.cloud`
     - Added a new example under [examples/backup-local](examples/backup-local) showing how to use the local backup feature
+      with `backup.cloud`
+    - Added a new property - `security.provisioner.passwordHash`
```
Comment (mihailradkov) [inline CHANGELOG.md:15]: Indentation is wrong

File: templates/graphdb/statefulset.yaml [hunk 4]
```diff
@@ -47,7 +47,7 @@ spec:
       spec:
         {{- toYaml .Values.import.volumeMount.volumeClaimTemplate.spec | nindent 8 }}
     {{- end }}
-    {{- if .Values.extraVolumeClaimTemplates }}
+    {{- with .Values.extraVolumeClaimTemplates }}
```
Comment (mihailradkov) [inline templates/graphdb/statefulset.yaml:50]: Remove this commit.

File: values.yaml [hunk 5]
```diff
@@ -258,6 +258,10 @@ security:
     # Note: The user is created regardless of if the existingSecret is provided or not.
     username: provisioner
     password: iHaveSuperpowers
+    # Define an initial password for the provisioner user.
```
Comment (mihailradkov) [inline values.yaml:261]:
```suggestion
    # Define an initial password for the provisioner user as a bcrypt hash.
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,8 +10,9 @@
     - Local backups support saving the GraphDB backup archives in volume from an existing persistent volume claim, configured
       with `backup.local`
     - Cloud backups support uploading the GraphDB backup archives in one of the supported cloud object storage services, configured
-      with `backup.cloud`
     - Added a new example under [examples/backup-local](examples/backup-local) showing how to use the local backup feature
+      with `backup.cloud`
+- Added a new property - `security.provisioner.passwordHash`
```
Comment (mihailradkov) [inline CHANGELOG.md:15]: It's not very helpful like this, at least add a short description of what it does/how it helps.

File: CHANGELOG.md [hunk 7]
```diff
@@ -10,8 +10,9 @@
     - Local backups support saving the GraphDB backup archives in volume from an existing persistent volume claim, configured
       with `backup.local`
     - Cloud backups support uploading the GraphDB backup archives in one of the supported cloud object storage services, configured
-      with `backup.cloud`
```
Comment (mihailradkov) [inline CHANGELOG.md:13]: Revert this change...

File: templates/_helpers.tpl [hunk 8]
```diff
@@ -97,3 +97,10 @@ Converts custom users YAML to a pretty JSON for insertion in users.js
   {{- end -}}
 {{- end -}}
 {{- end -}}
+
+{{/*
+Calculate provisoner's bcrypt-hashed password
+*/}}
+{{- define "graphdb.security.provisioner.passwordHash" -}}
+  {{- printf "{bcrypt}%s" ( htpasswd .Values.security.provisioner.username .Values.security.provisioner.password | trimPrefix (printf "%s:" .Values.security.provisioner.username)) -}}
+{{- end -}}
```
Comment (mihailradkov) [inline templates/_helpers.tpl:106]: Add final new line.

File: CHANGELOG.md [hunk 9]
```diff
@@ -10,8 +10,8 @@
...
-    - Added a new example under [examples/backup-local](examples/backup-local) showing how to use the local backup feature
+    - Added a new example under [examples/backup-local](examples/backup-local) showing how to use the local backup feature with `backup.cloud`
```
Comment (mihailradkov) [inline CHANGELOG.md:13]: > with `backup.cloud`

is relevant to the previous note...

**Review by mihailradkov — APPROVED** [review]

File: values.yaml [hunk 10]
```diff
@@ -258,6 +258,10 @@ security:
     # Note: The user is created regardless of if the existingSecret is provided or not.
     username: provisioner
     password: iHaveSuperpowers
+    # Define an initial password for the provisioner user as a bcrypt hash.
+    # If the password is not specified, it will use the default one specified in the provisioner.password.
+    # Important: This value must be a bcrypt encrypted hash, not a plaintext string. The format is "{bcrypt}<the-encrypted-hash>"
```
Comment (Secchol) [inline values.yaml:263]: Make the "{bcrypt}" automatically be added to the hash in the user.js file so the user only has to input the hash. This improves the user experience.
